[1단계 방탈출 결제 / 배포] 멍구(이유영) 미션 제출합니다.

yuyoungrhee opened the pull request:
안녕하세요 제이! 블랙잭에 이어 또 만나게 돼서 너무 반갑네요, 리뷰 잘 부탁드립니다!

### [이번 미션 범위 File changed만 확인](https://github.com/woowacourse/spring-roomescape-payment/pull/194/files/e27f514b8c48b5445a8175dad25a4fc8af1a4a7d..715bf48ab4e04a70032f72061a07fdf0dc373e0d)

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

  [결제/배포 프롤로그 링크](https://prolog.techcourse.co.kr/studylogs/4452)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

현재는 토스페이로 결제한다고 가정하고 로직이 토스페이에 강하게 결합되어 있습니다. 미션 진행하면서 다른 결제방식(카카오페이,핸드폰결제, 등)에 유연하게 설계하는 법을 고민해보려고 합니다.

### [이번 미션 범위 File changed만 확인](https://github.com/woowacourse/spring-roomescape-payment/pull/194/files/e27f514b8c48b5445a8175dad25a4fc8af1a4a7d..715bf48ab4e04a70032f72061a07fdf0dc373e0d)


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

Ryan-Dia committed:
chore: 이전 미션 세팅

Ryan-Dia committed:
feat: 클라이언트 코드 추가

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
feat: @Transactional 애너테이션 추가

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
feat: 예약 결제 기능 추가

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
feat: 결제 승인 예외 처리

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
feat: 결제 관련 예외 분리 - 사용자,개발자

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
fix: 불필요한 dto 제거

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
fix: secretKey 오타 수정

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
test: 결제 테스트 추

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
style: 로깅 추가 및 수정

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
feat: 결제 API timeout 설정

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
style: 메서드 이름 통일

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
style: 서비스레벨 맞게 메서드명 수정

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
style: 코드 컨벤션에 맞게 수정

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
refactor: 매직넘버 상수화

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
test: 테스트 port 관련 에러 수정

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
refactor: 상수를 환경변수로 빼기

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
fix: 응답값에 비밀번호 필드 제거

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
fix: type 변경

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
refactor: 사용하지 않는 메서드 제거

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
test: 테스트 메서드명 한글화

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
fix: transactional 애너테이션 추가

Co-authored-by: yuyoungrhee <bbclara@naver.com>

yuyoungrhee commented on src/main/java/roomescape/payment/exception/TossErrorResponse.java:
```diff
@@ -0,0 +1,11 @@
+public record TossErrorResponse(String code, String message, String data) {
+
+    public boolean isPaymentError() {
+        return code.equals("INVALID_API_KEY")
+                || code.equals("UNAUTHORIZED_KEY")
+                || code.equals("INCORRECT_BASIC_AUTH_FORMAT")
+                || code.equals("INVALID_AUTHORIZE_AUTH");
```
``` java
    @ExceptionHandler(HttpClientErrorException.class)
    public ResponseEntity<Object> handleBadRequestException(final HttpClientErrorException e, final WebRequest request)
    {
        final Optional<TossErrorResponse> response = extractResponse(e.getResponseBodyAsString());
        if (response.isEmpty() || response.get().isPaymentError()) {
            log.error(response.get().message(), e.getStatusCode());
            return buildResponseEntity(e, HttpStatus.INTERNAL_SERVER_ERROR, "결제 관련 내부 오류가 발생했습니다.", request);
        }
        return buildResponseEntity(e, e.getStatusCode(), response.get().message(), request);
    }

```

isPaymentError()를 사용하는 데가 여기밖에 없는데, 이미 Optional로 값 유무를 확인하고 있어서 그런일은 발생하지 않을 것 같아

Ryan-Dia committed:
style: final 키워드 추가

Co-authored-by: yuyoungrhee <bbclara@naver.com>

yuyoungrhee commented on src/test/java/roomescape/payment/controller/PaymentControllerTest.java:
```diff
@@ -0,0 +1,253 @@
+        });
+    }
+
+    @Test
+    void 타임아웃_시간외_응답_시_예외_처리() {
+        // given
+        wireMockServer.stubFor(post(urlEqualTo("/v1/payments/confirm"))
+                .willReturn(aResponse()
+                        .withStatus(200)
+                        .withHeader("Content-Type", "application/json")
+                        .withBody(createNormalBody())
+                        .withFixedDelay(5_000)
+                ));
```
withFixedDelay()를 통해 외부 API의 응답이 늦어지는 상황을 stub 했습니다.
timeout 설정에 대한 테스트인데, 이렇게 테스트할 경우 온전히 5초 이상을 기다려야하고 이런 외부 API 테스트가 늘어날 수록 부담이 되는 테스트코드라고 생각합니다. 외부 API에 대한 timeout 테스트는 하지 않는 것이 맞을까요? 이런 부분을 테스트하고 싶다면 어떻게 할 수 있을까요?

JunHoPark93 replied:
굉장히 좋은 질문이네요~
client 설정은 대부분 공유할거라서 (앱이 커져서 여러 client를 가질 수 있겠지만) 괜찮을것같은데요, 그래도 빠른 이터레이션을 돌리기에 부담이 된다고하면 ignore처리를 하면서 개발을 하는방법도 있을거고, 혹은 client bean의 설정값을 가져와 5초로 설정되어있는지 검증하는 방식으로 테스트를 해볼 수 있을것같아요.

yuyoungrhee replied:
감사합니다! 제가 어떤 것을 테스트하고 싶은지에 따라서 선택이 달라지겠네요. 

테스트 하고자 하는 것은 "timout설정이 5초로 잘 설정되었는가"가 아닌,
"**timeout 시 PaymentTimeoutException이 잘 터지는가**" 였기 때문에 테스트 yml 설정에서 timout 시간을 2초로 줄여서 2초 이내면 정상 응답, 2초 초과 시 예외 처리를 테스트하도록 했습니다. 이런 방식도 괜찮을까요?

[refactor: RestClient 설정을 공유하도록 리팩터링, 테스트 timout시간 축소](https://github.com/woowacourse/spring-roomescape-payment/pull/194/commits/715bf48ab4e04a70032f72061a07fdf0dc373e0d)

yuyoungrhee replied:
> client 설정은 대부분 공유할거라서 (앱이 커져서 여러 client를 가질 수 있겠지만) 괜찮을것같은데요, 

Client 설정을 공유한다면 timeout 설정을 공유할 것 같은데, 현재 구조로는 Client를 바로 TossRestClient를 빈으로 등록해버려서 설정을 공유하기가 어려운 구조였습니다. 그래서 해당 코멘트에서 힌트를 얻고 RestClient 설정과 PaymentRestClient 설정을 분리했어요.

기존
``` java
    @Bean
    public TossRestClient tossRestClient(final Builder restClientBuilder, final RestClientProperties restClientProperties, final TossPaymentProperties tossPaymentProperties) {
        connectTimeout = restClientProperties.getConnectTimeout();
        connectionRequestTimeout = restClientProperties.getConnectionRequestTimeout();
        readTimeout = restClientProperties.getReadTimeout();
        return new TossRestClient(restClientBuilder
                .baseUrl(restClientProperties.getBaseUrl())
                .requestFactory(clientHttpRequestFactory())
                .build(),
                tossPaymentProperties
        );
    }
```

변경
``` java

// RestClientConfig.java
    @Bean
    public RestClient.Builder restClientBuilder(final RestClientProperties restClientProperties) {
        connectTimeout = restClientProperties.getConnectTimeout();
        connectionRequestTimeout = restClientProperties.getConnectionRequestTimeout();
        readTimeout = restClientProperties.getReadTimeout();
        return RestClient.builder()
                .requestFactory(clientHttpRequestFactory());
    }

//PaymentRestClientConfig.java
    @Bean
    public TossRestClient tossRestClient(RestClient.Builder restClientBuilder, TossPaymentProperties tossPaymentProperties) {

        return new TossRestClient(restClientBuilder
                .baseUrl(tossPaymentProperties.getBaseUrl())
                .build(),
                tossPaymentProperties
        );
    }
```

JunHoPark93 replied:
> "timeout 시 PaymentTimeoutException이 잘 터지는가" 였기 때문에 테스트 yml 설정에서 timout 시간을 2초로 줄여서 2초 이내면 정상 응답, 2초 초과 시 예외 처리를 테스트하도록 했습니다. 이런 방식도 괜찮을까요?

넵 실제로 보려면 말씀하신 방식으로 해야될거고 테스트시간이 길어지는 것에 대한우려로 그 시간을 2초로 줄이신다는것 같네요. 시간은 줄긴했지만 그래도 지연이 있는것은 동일하기때문에 동작확인정도만 해보고 ignore처리후 설정값이 잘 세팅되었는지 테스트코드로 검증하는것도 나쁘진 않을것같네요~

Ryan-Dia committed:
fix: jwt 인코딩 설정

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
fix: NPE 발생 가능성 있는 코드 수정

Co-authored-by: yuyoungrhee <bbclara@naver.com>

Ryan-Dia committed:
refactor: widgetSecretKey properties로 분리

Co-authored-by: yuyoungrhee <bbclara@naver.com>

yuyoungrhee reviewed (commented):
테스트에 관한 간단한 질문 하나 남깁니다!

Ryan-Dia committed:
style: 불필요한 주석 및 공백 제거

Co-authored-by: yuyoungrhee <bbclara@naver.com>

JunHoPark93 commented on src/main/java/roomescape/payment/service/PaymentService.java:
```diff
@@ -0,0 +1,41 @@
+        final Reservation reservation = reservationService.findById(reservationResponse.id());
+        log.debug("Reservation ID: {}", reservation.getId());
+
+        final Payment payment = Payment.builder()
```
Payment생성을 분리하여 메서드 길이를 줄여보면 어떨까요?

yuyoungrhee replied:
builder를 사용하니 가독성이 좀 떨어지네요. 메서드로 분리했습니다!
[refactor(Payment): 메서드 분리로 가독성 향상](https://github.com/woowacourse/spring-roomescape-payment/pull/194/commits/b11e42f4780a8f1494dbda1edecbbb35236e8c00)

JunHoPark93 commented on src/main/java/roomescape/payment/controller/PaymentController.java:
```diff
@@ -0,0 +1,41 @@
+                request.orderId(),
+                request.amount()
+        );
+        return tossRestClient.confirm(tossPaymentRequest);
```
외부호출이후에 데이터베이스에 저장되는 내용은 없어도되는걸까요?

yuyoungrhee replied:
DB 저장 후 외부 호출하도록 변경했습니다!

[refactor: 결제 승인 & 결제 정보 저장 로직을 controller&service에 재분배](https://github.com/woowacourse/spring-roomescape-payment/pull/194/commits/449d19f5cdeefc0b1b371b301dc8c75c10cfb0a4)

``` java
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/confirm/tossPay")
    public void confirmPayment(@RequestBody @Valid final ReservationPaymentRequest request, final LoginMember loginMember) {
        final TossPaymentRequest tossPaymentRequest = new TossPaymentRequest(
                request.paymentKey(),
                request.orderId(),
                request.amount()
        );
        paymentService.saveReservationPayment(request, loginMember);
        paymentService.confirmPayment(tossPaymentRequest);
    }
```

JunHoPark93 replied:
컨트롤러내에서 로직을 전개하기보다는 paymentService안으로 넣어주시는게 좋아보입니다.

JunHoPark93 commented on src/main/java/roomescape/payment/domain/Payment.java:
```diff
@@ -0,0 +1,60 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @JoinColumn(nullable = false)
+    @ManyToOne(optional = false, fetch = FetchType.LAZY)
+    private Member member;
+
+    @JoinColumn(nullable = false)
+    @OneToOne(optional = false, fetch = FetchType.LAZY)
+    private Reservation reservation;
```
결제에 member와 reservation을 다 가지게 만들어주셨네요~
설계의도를 말씀주실 수 있으실까요?

yuyoungrhee replied:
결제를 정의할 때 필요한 것은 <**결제 정보**(paymentKey, orderId, amount 등등), **결제한 사람**(Member), **어떤 예약에 대한 결제인지**(Reservation)> 라고 생각했습니다.
따라서 이 세 가지를 모두 가지고 있어야 Payment라는 객체를 표현할 수 있겠다고 생각하여 이렇게 설계했습니다.

다른 후보
- **Reservation -> Payment로, Reservation이 OneToOne연관관계의 주인**
결제라는 정보는 민감정보라고 할 수 있는데, 다른 객체를 통해 Payment에 접근할 수 있는 가능성을 막고 싶었습니다.

- **Member -> Payment로, Member가 OneToMany로 List<Payment>를 가지고 있음**
Member 엔티티의 조회나 기능에서는 결제(Payment)가 항상 필요한 것은 아닙니다. 내 정보 페이지를 볼 때는 결제 정보가 필요 없지만, 결제 내역을 조회하는 기능에서는 Payment가 필요하게 됩니다. 반면, **Payment 객체는 반드시 어떤 회원(Member)과 연결되어 있어야 그 의미가 완성됩니다.** 결제를 조회하거나 다룰 때는 언제나 그 결제가 누구의 것인지를 알아야 하기 때문입니다. 따라서 Member가 Payment를 알 필요는 없다고 생각했습니다.

JunHoPark93 commented on src/main/java/roomescape/payment/domain/Payment.java:
```diff
@@ -0,0 +1,60 @@
+    @OneToOne(optional = false, fetch = FetchType.LAZY)
+    private Reservation reservation;
+
+    @Column(nullable = false)
+    private String paymentKey;
+
+    @Column(nullable = false)
+    private String orderId;
```
고유한 값들이 있을까요?

yuyoungrhee replied:
그러네요! payemtnKey와 orderId는 고유한 값입니다.
orderId: 내 서버에서 부여하는 고유하게 생성하는 값으로, 현재 프론트 코드에서 생성하고 있습니다.
paymentKey: 결제사인 토스에서 부여하는 키로, 결제마다 고유합니다. 이 키를 이용해서 결제 취소 등의 소통을 할 수 있습니다.

따라서 paymentKey와 orderId는 unique속성을 DB에게도 부여해야겠습니다. 리팩터링하며 반영했어요.

``` java
    @Column(nullable = false, unique = true)
    private String paymentKey;

    @Column(nullable = false, unique = true)
    private String orderId;
```

JunHoPark93 commented on src/main/java/roomescape/payment/infrastructure/RestClientConfig.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Bean
+    public TossRestClient tossRestClient(final Builder restClientBuilder, final RestClientProperties restClientProperties, final TossPaymentProperties tossPaymentProperties) {
+        connectTimeout = restClientProperties.getConnectTimeout();
+        connectionRequestTimeout = restClientProperties.getConnectionRequestTimeout();
+        readTimeout = restClientProperties.getReadTimeout();
+        return new TossRestClient(restClientBuilder
+                .baseUrl(restClientProperties.getBaseUrl())
+                .requestFactory(clientHttpRequestFactory())
+                .build(), tossPaymentProperties.getWidgetSecretKey());
```
client 설정값을 다뤄보셨네요~
각각 어떤 역할을 하는지 남겨주시면 좋을것같아요.

yuyoungrhee replied:
**connectTimeout**
클라이언트가 서버에 연결을 시도할 때, 연결이 성공하기까지 기다리는 최대 시간
- 설정값: 2초-5초
  connect 자체가 안 되는 상황은 빨리 실패시키는 것이 좋음. (ex. 서버가 죽었거나 포트가 잘못되었을 경우)

**connectionRequestTimeout**
커넥션 풀에서 커넥션을 얻어오기까지의 최대 대기 시간
- 설정값: 1-2초
  커넥션 풀이 부족한 경우 내서버 과부하이므로, 풀 대기가 길어지지 않도록 빨리 실패시키는 것이 좋음.

**readTimeout**
요청을 서버로 보내고, 서버 응답 데이터를 받기까지의 최대 대기 시간
- 설정값: 5초-30초
  서버가 정상 작동 중이라도 처리 시간이 길어질 경우. API 종류에 따라 여유를 둬야한다(결제 API의 경우 처리 시간 자체가 길 수 있음)

JunHoPark93 replied:
내역을 잘 찾아봐주셨네요~
설정값 범위는 통상적인 설정 범위를 찾아주신걸까요? 

추가적으로 현재 이 앱에 적당한 설정값은 얼마일지 추정해서 남겨주셔도 좋을것같아요~

JunHoPark93 commented on src/main/java/roomescape/payment/infrastructure/TossRestClient.java:
```diff
@@ -0,0 +1,46 @@
+                    .retrieve()
+                    .body(TossPaymentResponse.class);
+        } catch (ResourceAccessException ex) {
+            if (ex.getCause() instanceof SocketTimeoutException) {
+                log.error("토스 결제 confirm 요청 타임아웃", ex);
+                throw new PaymentTimeoutException("결제 시스템이 응답하지 않아 시간이 초과되었습니다.");
+            }
```
SocketTimeoutException만 명시적으로 처리하신 이유가 있으실까요?

yuyoungrhee replied:
타임아웃은 “외부 시스템(토스)에서 실제로 요청을 처리했을 가능성”이 존재하기 때문입니다.

ResourceAccessException 중에서, 실제로 외부 서버에 요청이 성공했을 가능성이 있는 경우는 SocketTimeoutException뿐입니다. 다른 예외로는 
- ConnectException(연결 자체가 실패), 
- UnknownHostException(호스트 이름을 찾을 수 없음)
- InterruptedIOException(I/O 중 쓰레드 인터럽트 등으로 중단됨)

등등, 더 많지만 전반적으로 외부서버까지 도달하지 못하는 경우에 ResourceAccessException이 발생합니다. 따라서 타임아웃만 “서버가 실제로 처리했을 수 있다”를 고려해야합니다.

때문에 timeout 시 다시 토스에 결제 확인 API를 보낸 후 결제가 되었다면 결제 취소 API를 호출하는 과정까지 구현해야하나, 이번 미션의 학습 범위는 아닌 것 같아 구현하지 않고 주석으로 남겨두었습니다.

[chore: timeout 구현 필요 부분 주석 추가](https://github.com/woowacourse/spring-roomescape-payment/pull/194/commits/20f6f2d112d960dbc0f58ec791d965108e3780fd)

JunHoPark93 commented on src/main/java/roomescape/payment/infrastructure/TossRestClient.java:
```diff
@@ -0,0 +1,46 @@
+@Slf4j
+public class TossRestClient {
+
+    private final RestClient restClient;
```
RestClient를 선택하신 이유가 있을까요?

yuyoungrhee replied:
Spring에서 지원하는 클라이언트의 경우 RestTmplate, WebClient, RestClient입니다. 자바 자체의 HttpClient, 외부 라이브러리 Client도 존재하나 스프링을 사용하고 있는 입장에서 3개를 먼저 고려했습니다.

여기서 RestTemplate은 점차 레거시가 되어가고 있고 스프링에서도 RestClient 사용을 권장하므로 RestTemplate도 고려하지 않았습니다. 실제로 학습 테스트에서도 RestClient가 가독성이 더 좋고 예외처리가 간편했습니다. 

여기서 그럼 WebClient(논블로킹) vs RestClient(동기) 를 비교해야합니다. 성능 면에서는 WebClient가 좋습니다. 쓰레드 낭비를 줄이며, 트래픽이 대규모라면 효율적입니다. 다만 Spring MVC가 아닌 Spring WebFlux에서 제공하고 있기 때문에 학습곡선이 있는 편이고, 예외처리, 기존 MVC 코드와 호환되도록 매핑하는 과정 등 코드 복잡도가 올라갑니다. 또한 요청이 많지 않은 경우 RestClient와 성능 차이는 미미합니다.

결론적으로 현재 서비스가 대규모 트래픽을 받지 않고, 작은 서비스이기 때문에 RestClient로도 충분하다고 판단했습니다.

JunHoPark93 commented on src/main/java/roomescape/payment/infrastructure/TossRestClient.java:
```diff
@@ -0,0 +1,46 @@
+
+@RequiredArgsConstructor
+@Slf4j
+public class TossRestClient {
```
외부 호출 클라이언트를 정의해주셨네요~
테스트코드는 어떻게 작성해보면 좋을까요?

yuyoungrhee replied:
WireMockServer를 이용해서 작성했습니다. 
[TossRestClientTest.java](https://github.com/woowacourse/spring-roomescape-payment/pull/194/files/e27f514b8c48b5445a8175dad25a4fc8af1a4a7d..715bf48ab4e04a70032f72061a07fdf0dc373e0d#diff-338e47406fb627e621df7cf01fd639a325b8a27b4c3153bfb77b9ffd0daf38ec)

### 외부 API 연동 테스트 선택지
1) Mocking
외부 API를 호출하는 객체 (RestClient)를 mock으로 만들고, 예상 응답을 리턴하게 설정 
2) HTTP Stub 서버 사용 (예: WireMock, MockServer)
외부 API를 흉내내는 Stub 서버를 띄우고, HTTP 요청/응답을 가로채서 예측 가능한 응답을 반환 
3) E2E
실제 API를 호출하는 테스트 

### 결정 과정
- 테스트 하고 싶은 것: RestClient가 제대로 작동하는지 였기 때문에 1은 제외. (1은 RestClient 자체를 mocking하는 방법이기 때문에)
- 테스트 안전성: E2E는 불가, 실제 API는 API 서버 문제로 잠깐 문제가 생긴다면 테스트가 실패함. 내가 짠 코드 외의 요인에 의해 실패하는 테스트는 테스트로서 의미가 없다고 생각. 

따라서 2번 Http Stub을 선택했습니다.

### Http Stub 테스트의 장단점
- 장점
  - 실제 API와 유사한 환경에서 테스트 가능
  - 외부 API의 예측 가능한 응답 시뮬레이션 가능(응답 코드, 반환 body 등 모든 상황 stub하여 내 코드에서 어떻게 반응해야하는지 시뮬레이션 가능)
  - 외부 API에 의존하지 않는 독립적인 테스트와 안전성 확보 
- 단점
  - 실제 API가 변경되면 테스트가 무의미해짐-실제 응답 자체를 stub하므로
  - 요청/응답을 미리 정의해두어야 함-API가 많아질수록 Stub 관리가 복잡해짐
  - 실제 환경과 다를 수 있음-E2E 테스트나 실제 프로덕션에서 테스트하는 것만큼의 신뢰도는 어려움

JunHoPark93 replied:
정리잘해주셨네요!

JunHoPark93 reviewed (changes requested):
안녕하세요 멍구~
1단계 구현 고생많으셨습니다.
몇가지 코멘트 남겨두었는데요, 확인 후 의견 부탁드릴게요~

yuyoungrhee committed:
refactor: view를 위한 enum 위치 이동 및 클래스명 변경

- ReservationStatus가 예약인지 대기인지 판별하는 것은 view를 위한 enum이었음

yuyoungrhee committed:
refactor: 결제 승인 & 결제 정보 저장 로직을 controller&service에 재분배

- service: client를 사용하여 toss API 호출 및 결제 정보 업데이트
- controller: 예약&결제 정보 저장 및 결제 승인

yuyoungrhee committed:
fix: 에러 파싱이 되지 않는 오류 수정

yuyoungrhee committed:
fix(Payment): amount unique 속성 삭제

yuyoungrhee committed:
refactor(Payment): 메서드 분리로 가독성 향상

yuyoungrhee committed:
chore: timeout 구현 필요 부분 주석 추가

yuyoungrhee committed:
test(TossRestClient): 테스트 케이스 추가

yuyoungrhee committed:
refactor: RestClient 설정을 공유하도록 리팩터링, 테스트 timout시간 축소

yuyoungrhee commented:
지노 리뷰 감사합니다! 혼자 해결하기 어려운 부분들에 대해서 힌트 많이 얻었습니다.
[이번 리팩터링 변경 사항 File changed](https://github.com/woowacourse/spring-roomescape-payment/pull/194/files/6a8e43547766d53eb58f78e988cf98a92687ef33..715bf48ab4e04a70032f72061a07fdf0dc373e0d)

**변경 사항 정리**
- tossRestClient를 Controller가 아닌 Service에서 사용하도록 함
  - 외부 API 호출은 비즈니스 로직을 처리하기 위한 일부 과정이므로 Controller가 아닌 Service에 위치해야함

- 결제 승인&DB 저장 순서 및 로직 변경
  - 기존: toss API로 confirm 요청 -> 성공 시 DB 저장 (프론트에서 각각 API를 호출하고 있었습니다)
  - 변경: 결제 전 DB 저장(결제 중 상태) -> toss API로 confirm 요청 -> 결제 성공/실패에 따라 결제상태 update(결제 완료/결제 실패)
 
- 예외를 TossRestClient에서 일부 포장하여 던짐
  - RestClient의 onStatus를 활용하여 응답 결과에 따라 4xx, 5xx 에러를 커스텀 예외로 포장
  - 여기서 던진 예외는 PaymentHandler에서 형식에 맞게 최종 처리

yuyoungrhee committed:
fix: 로그인 만료시 500에러 뜨는 문제 해결

- Filter에서 UnauthorizedException도 catch하도록 수정

JunHoPark93 reviewed (approved):
안녕하세요 멍구~
리뷰 반영 고생많으셨습니다.
찾아볼 내용들이 많았는데 잘 해주셨습니다!
몇가지 추가 반영 사항은 다음 단계에서 진행주셔도 될것같네요~
